src/core/components/pin.js

```diff
@@ -0,0 +1,563 @@
+        })
+      }, (err) => {
+        if (err) { return callback(err) }
+        callback(null, indirectKeys.toStringArray())
+      })
+    },
+
+    internalKeys: () => {
+      return internalPins.toArray()
+    },
+
+    internalKeyStrings: () => {
+      return internalPins.toStringArray()
+    },
+
+    // encodes and writes pin key sets to the datastore
+    // each key set will be stored as a DAG node, and a root node will link to both
+    flush: promisify((callback) => {
+      // callback (err, root)
+      const newInternalPins = new KeySet()
+      const logInternalKey = (mh) => newInternalPins.add(mh)
+      const handle = {
+        put: (k, v, cb) => {
+          handle[k] = v
+          cb()
+        }
+      }
+      waterfall([
+        // create link to direct keys node
+        (cb) => pin.set.storeSet(pin.directKeys(), logInternalKey, cb),
+        (dRoot, cb) => DAGLink.create(pin.types.direct, dRoot.size, dRoot.multihash, cb),
+        (dLink, cb) => handle.put('dLink', dLink, cb),
+        // create link to recursive keys node
+        (cb) => pin.set.storeSet(pin.recursiveKeys(), logInternalKey, cb),
+        (rRoot, cb) => DAGLink.create(pin.types.recursive, rRoot.size, rRoot.multihash, cb),
+        (rLink, cb) => handle.put('rLink', rLink, cb),
+        // the pin-set nodes link to an empty node, so make sure it's added to dag
+        (cb) => DAGNode.create(new Buffer(0), cb),
+        (empty, cb) => dag.put(empty, {cid: new CID(empty.multihash)}, cb),
+        // create root node with links to direct and recursive nodes
+        (cid, cb) => DAGNode.create(new Buffer(0), [handle.dLink, handle.rLink], cb),
+        (root, cb) => handle.put('root', root, cb),
+        // add the root node to dag
+        (cb) => dag.put(handle.root, {cid: new CID(handle.root.multihash)}, cb),
+        // update the internal pin set
+        (cid, cb) => cb(null, logInternalKey(handle.root.multihash)),
+        // save serialized root to datastore under a consistent key
+        (_, cb) => repo.closed ? repo.datastore.open(cb) : cb(null, null),  // temp
+        (_, cb) => repo.datastore.put(pinDataStoreKey, handle.root.serialized, cb)
+      ], (err, result) => {
+        if (err) { return callback(err) }
+        internalPins = newInternalPins
+        return callback(null, handle.root)
+      })
+    }),
+
+    load: promisify((callback) => {
+      // callback (err)
+      const newInternalPins = new KeySet()
+      const logInternalKey = (mh) => newInternalPins.add(mh)
+      const handle = {
+        put: (k, v, cb) => {
+          handle[k] = v
+          cb()
+        }
+      }
+      waterfall([
+        (cb) => repo.closed ? repo.datastore.open(cb) : cb(null, null),  // temp
+        (_, cb) => repo.datastore.has(pinDataStoreKey, cb),
+        (has, cb) => has ? cb() : cb('break'),
+        (cb) => repo.datastore.get(pinDataStoreKey, cb),
+        (serialized, cb) => dagPB.util.deserialize(serialized, cb),
```

Comment:
The PinSet object is a dag Protobuf? I believe this should be `go-ipfs/pin/internal/pb/header.proto`